drivers/usb/device/usb_dc_mcux.c

```diff
@@ -68,7 +71,35 @@ static void usb_isr_handler(void);
 #define EP_ABS_IDX(ep)		(USB_EP_GET_IDX(ep) * 2 + \
 					(USB_EP_GET_DIR(ep) >> 7))
 #define NUM_OF_EP_MAX		(DT_INST_PROP(0, num_bidir_endpoints) * 2)
-#define CONTROLLER_ID		(DT_INST_ENUM_IDX(0, usb_controller_index))
+
+#define COUNT_INST(n) 1 +
+#define NUM_INSTS (DT_INST_FOREACH_STATUS_OKAY(COUNT_INST) 0)
+BUILD_ASSERT(NUM_INSTS <= 1, "Only one USB device supported");
+
+/* Controller ID is for HAL usage */
+#if defined(CONFIG_SOC_SERIES_IMX_RT5XX) || \
```

Comment:
For this SOC specific section- could we use an approach like this?

```
/* LPC3511 IP */
#ifdef CONFIG_USB_DC_NXP_LPCIP3511
#if DT_INST_ENUM_HAS_VALUE(0, maximum_speed, high_speed)
#define CONTROLLER_ID kUSB_ControllerLpcIp3511Hs0
#else
#define CONTROLLER_ID kUSB_ControllerLpcIp3511Hs0
#endif

/* EHCI IP */
#ifdef CONFIG_USB_DC_NXP_EHCI
#define CONTROLLER_ID kUSB_ControllerEhci##DT_INST_PROP(0, index)
#endif
```
Not sure what your thoughts on this are

Reply:
Introducing an index property defeats the purpose of solving the issue linked since we are once again putting things in DT to satisfy HAL enum names. Also we had a similar discussion on S32 PRs before and decided to avoid putting index numbers in DT to satisfy HAL arrays.

Reply:
I don't like the solution, but I see a way we could avoid the SOC specific Kconfig definitions here without an index property- since the `CONTROLLER_ID` value is used at runtime, we could define a function to get the controller ID based on the base address of the controller.

Something like this:
```
/* Controller ID is for HAL usage */
/* LPC3511 IP */
#ifdef CONFIG_USB_DC_NXP_LPCIP3511
static usb_controller_index_t mcux_usb_controller_enum(void)
{
        return DT_INST_ENUM_HAS_VALUE(0, maximum_speed, high_speed) ?
                kUSB_ControllerLpcIp3511Hs0 :
                kUSB_ControllerLpcIp3511Fs0
}
#endif

/* EHCI IP */
#ifdef CONFIG_USB_DC_NXP_EHCI
static usb_controller_index_t mcux_usb_controller_enum(void)
{
/* This section is pulled from the MCUX SDK HAL driver for the ECHI, so the USBHS_BASE_ADDRS definition should
 * exist for any SOC with ECHI IP
 */
#if defined(USBHS_STACK_BASE_ADDRS)
        uint32_t base_addrs[] = USBHS_STACK_BASE_ADDRS;
#else
        uint32_t base_addrs[] = USBHS_BASE_ADDRS;
#endif
        return DT_INST_REG_ADDR(0) == base_addrs[0] ?
                kUSB_ControllerEhci0 :
                kUSB_ControllerEhci1;
}
#endif
```

We would then call `mcux_usb_controller_enum` in the `usb_dc_attach` function to get the controller ID value.

Reply:
It is a good idea but there is no guarantee that this pattern holds for future SOCs so I don't see the advantage of making this over-complicated. I would rather maintain just this list of things per SOC rather than do these gymnastics with every new great idea for an api change the HAL comes up with for every new release. For example, there could be an SOC with two instances of LPC3511 which this doesn't account for because there might be a case where only the index 1 is enabled. And there could be an SOC where the HAL adds yet another random name of the array like they have already done with USBHS_STACK_BASE_ADDRS vs USBHS_BASE_ADDRS. And sometimes the HAL puts a random 0 in the first index of USB_BASE_ADDRS which would break things. I don't think having the SOC ifdefs is really adding any more technical debt than this PR is solving, someone enabling an SOC already currently has to figure out which hal enum corresponds to the modules on the chip, the only difference is now putting it in this driver file instead of the devicetree.

If we really want a better solution it has to come from the SDK itself is my opinion. Either that or make the driver zephyr native as the usb_dc_kinetis is, but both of those things are outside the scope of this PR which is just a quick fix of the abuse of devicetree.

Reply:
> It is a good idea but there is no guarantee that this pattern holds for future SOCs so I don't see the advantage of making this over-complicated

I expect this pattern should hold for new SOCs unless the HAL driver itself is modified. The code in the ECHI HAL driver for determining the base address to use for the USB IP looks like this:
```
#if defined(USBHS_STACK_BASE_ADDRS)
    uint32_t ehci_base[] = USBHS_STACK_BASE_ADDRS;
#else
    uint32_t ehci_base[] = USBHS_BASE_ADDRS;
#endif
....
ehciState->registerBase = (USBHS_Type *)ehci_base[controllerId - (uint8_t)kUSB_ControllerEhci0];
```

So we *should* always be able to determine the correct controller ID from the base address with something like this:

```
#if defined(USBHS_STACK_BASE_ADDRS)
        uint32_t base_addrs[] = USBHS_STACK_BASE_ADDRS;
#else
        uint32_t base_addrs[] = USBHS_BASE_ADDRS;
#endif
for (uint8_t i = 0; i < ARRAY_SIZE(base_addrs); i++) {
	if (base_addrs[i] == DT_INST_REG_ADDR(0)) {
		return kUSB_ControllerEhci0 + i;
}
```

> And sometimes the HAL puts a random 0 in the first index of USB_BASE_ADDRS which would break things. 

If you look at the RT1052, which does this- the `USBHS_STACK_BASE_ADDRS` macro is also defined with the correct array for the ECHI driver. So I'm not sure this would be an issue, since we'd just use the `USBHS_STACK_BASE_ADDRS` definition.

The core thing I want to avoid here is a growing ladder of `#ifdef` blocks for each SOC NXP enables with this driver. Ideally a developer should be able to determine the type of IP block an SOC is using for USB, and the driver should compile and work with it.

Reply:
I have said why I disagree and my opinion has not changed. Additionally, you are proposing runtime overhead for something that can be determined at compile time. And as you are pointing out, trying to make this more and more generic just leads to more and more runtime overhead. I don't think there is any point to attempt to optimize this here with these complications when we know the most optimal fix is going to come from the SDK and remove the need from any of this. I'm not changing this in the PR.

Reply:
Understood. At the end of the day, this fixes the issue. I don't particularly like the approach, but I'm not going to block over it.